src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,518 @@ extern "C" {
+//	@function:
+//		CQueryMutators::GetGroupUniqueTleReferencesWalker
+//
+//	@doc:
+//		Goes over the groupClause expression tree and collects unique
+//		references to targetlist entries.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GetGroupUniqueTleReferencesWalker(
+	Node *node, SContexGetGroupUniqueTleWalker *context)
+{
+	if (NULL == node)
+	{
+		return false;
+	}
+
+	if (IsA(node, SortGroupClause))
+	{
+		SortGroupClause *sgc = (SortGroupClause *) node;
+		context->m_grouping_tle_refs = gpdb::LAppendUniqueInt(
+			context->m_grouping_tle_refs, sgc->tleSortGroupRef);
+		return false;
+	}
+	return gpdb::WalkExpressionTree(
+		node, (ExprWalkerFn) GetGroupUniqueTleReferencesWalker, context);
+};
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::GroupingFuncRewriteWalker
+//
+//	@doc:
+//		Goes over the expression tree and update arguments of GROUPING
+//		function.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GroupingFuncRewriteWalker(
+	Node *node, SContexGroupingFuncRewriteWalker *context)
+{
+	if (NULL == node)
+	{
+		return false;
+	}
+
+	if (IsA(node, GroupingFunc))
+	{
+		GroupingFunc *gf = (GroupingFunc *) node;
+		ListCell *arg_lc;
+		List *newargs = NIL;
+
+		gf->ngrpcols = context->m_ngrpcols;
+
+		// For each argument in gf->args, find its new position
+		ForEach(arg_lc, gf->args)
+		{
+			int arg = intVal(lfirst(arg_lc));
+			int new_arg =
+				gpdb::ListNthInt(context->m_grouping_tle_refs_mapping, arg);
+			newargs = gpdb::LAppend(newargs, gpdb::MakeIntegerValue(new_arg));
+		}
+
+		// Free gf->args since we do not need it any more.
+		gpdb::ListFreeDeep(gf->args);
+		gf->args = newargs;
+		return false;
+	}
+	return gpdb::WalkExpressionTree(
+		node, (ExprWalkerFn) GroupingFuncRewriteWalker, context);
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::GetVarsWithoutTleWalker
+//
+//	@doc:
+//		Goes over the expression tree, finds all vars that do not have relevant
+//		targetlist entries, and adds resjunk targetlist entries for such vars
+//		to a m_tlist_addition in the context, that later will be added to the
+//		original target list.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GetVarsWithoutTleWalker(
+	Node *node, SContextFixGroupDependentTargets *context)
+{
+	if (NULL == node)
+	{
+		return false;
+	}
+
+	if (IsA(node, Aggref))
+	{
+		// do not examine what is inside Aggref
+		return false;
+	}
+
+	// find all vars that do not have relevant targetlist entries, and add
+	// resjunk targetlist entries for such vars to m_tlist_addition
+	if (IsA(node, Var))
+	{
+		bool found = false;
+		Var *var = (Var *) gpdb::CopyObject(node);
+		if (var->varlevelsup == context->m_current_query_level)
+		{
+			var->varlevelsup = 0;
```

Comment:
Is that guaranteed that all vars of m_query->targetlist point  to the m_query rtes?

Reply:
Possibly we should add a check for var->varlevelsup == 0 before line 449 where we fetch rte for target entry:
```
					RangeTblEntry *rte = (RangeTblEntry *) gpdb::ListNth(
						query->rtable, var->varno - 1);
```
I'm not sure if this situation is possible, but to to be on a safe side I guess better to add such check.

Reply:
Added this check